File: mgen-cpplib/src/test/cpp/src/tests/ExtremeValues.cpp
#include "TestBase.h"

#include "gameworld/types/ClassRegistry.h"
#include "mgen/serialization/BinaryWriter.h"
#include "mgen/serialization/BinaryReader.h"
#include "mgen/serialization/VectorInputStream.h"
#include "mgen/serialization/VectorOutputStream.h"
#include "mgen/serialization/JsonPrettyWriter.h"
#include "mgen/serialization/JsonWriter.h"
#include "mgen/serialization/JsonReader.h"

#define FLOAT_APPRIX_EQ(a,b) ( fabs((a) - (b)) <= fabs(a * 1e-5))

/////////////////////////////////////////////////////////////////////

using namespace gameworld::types;
using namespace gameworld::types::basemodule1;

class ExtremeValuesTestData {
public:

    typedef mgen::VectorOutputStream OS;
    typedef mgen::VectorInputStream IS;
    typedef ClassRegistry CR;

    CR registry;
    std::vector<char> buffer;
    OS out;
    IS in;

    mgen::JsonWriter<OS, CR> jsonWriter;
    mgen::JsonWriter<OS, CR> jsonWriterCompact;
    mgen::JsonPrettyWriter<OS, CR> jsonPrettyWriter;
    mgen::JsonPrettyWriter<OS, CR> jsonPrettyWriterCompact;
    mgen::BinaryWriter<OS, CR> binaryWriter;
    mgen::BinaryWriter<OS, CR> binaryWriterCompact;

    mgen::JsonReader<IS, CR> jsonReader;
    mgen::BinaryReader<IS, CR> binaryReader;

    ExtremeValuesTestData() :
                    out(buffer),
                    in(buffer),
                    jsonWriter(out, registry),
                    jsonWriterCompact(out, registry, true),
                    jsonPrettyWriter(out, registry),
                    jsonPrettyWriterCompact(out, registry, true),
                    binaryWriter(out, registry),
                    binaryWriterCompact(out, registry, true),
                    jsonReader(in, registry),
                    binaryReader(in, registry) {
    }

    void reset() {
        out.reset();
        in.reset();
    }

};

#define FOREACH_SERIALIZER(f) { \
    f(testData, testData.jsonWriter, testData.jsonReader); \
    f(testData, testData.jsonWriterCompact, testData.jsonReader); \
    f(testData, testData.jsonPrettyWriter, testData.jsonReader); \
    f(testData, testData.jsonPrettyWriterCompact, testData.jsonReader); \
    f(testData, testData.binaryWriter, testData.binaryReader); \
    f(testData, testData.binaryWriterCompact, testData.binaryReader); \
}

BEGIN_TEST_GROUP(ExtremeValues)

/////////////////////////////////////////////////////////////////////

static int s_nSerializers = 0;

template<typename WriterType, typename ReaderType>
static void haveSerializers(ExtremeValuesTestData& testData, WriterType& writer, ReaderType& reader) {
    s_nSerializers++;
}

BEGIN_TEST("Have serializers")

    ExtremeValuesTestData testData;
    ClassRegistry& registry = testData.registry;

    FOREACH_SERIALIZER(haveSerializers);

    ASSERT(s_nSerializers >= 2);

END_TEST

/////////////////////////////////////////////////////////////////////

template<typename WriterType, typename ReaderType>
static void testFixedPoint(ExtremeValuesTestData& testData, WriterType& writer, ReaderType& reader) {

    Car car, carBack;
    car._setAllFieldsSet(true, mgen::DEEP);

    car.setId(LONG_LONG_MAX);
    writer.writeObject(car);
    carBack = reader.template readStatic<Car>();
    ASSERT(carBack.getId() == LONG_LONG_MAX);
    testData.reset();

    car.setId(LONG_LONG_MIN);
    writer.writeObject(car);
    carBack = reader.template readStatic<Car>();
    ASSERT(carBack.getId() == LONG_LONG_MIN);
    testData.reset();

    car.setId(0);
    writer.writeObject(car);
    carBack = reader.template readStatic<Car>();
    ASSERT(carBack.getId() == 0);
    testData.reset();

    car.setId(1);
    writer.writeObject(car);
    carBack = reader.template readStatic<Car>();
    ASSERT(carBack.getId() == 1);
    testData.reset();

    car.setId(-1);
    writer.writeObject(car);
    carBack = reader.template readStatic<Car>();
    ASSERT(carBack.getId() == -1);
    testData.reset();

    car.setTopSpeed(INT_MAX);
    writer.writeObject(car);
    carBack = reader.template readStatic<Car>();
    ASSERT(carBack.getTopSpeed() == INT_MAX);
    testData.reset();

    car.setTopSpeed(INT_MIN);
    writer.writeObject(car);
    carBack = reader.template readStatic<Car>();
    ASSERT(carBack.getTopSpeed() == INT_MIN);
    testData.reset();

    car.setTopSpeed(0);
    writer.writeObject(car);
    carBack = reader.template readStatic<Car>();
    ASSERT(carBack.getTopSpeed() == 0);
    testData.reset();

    car.setTopSpeed(1);
    writer.writeObject(car);
    carBack = reader.template readStatic<Car>();
    ASSERT(carBack.getTopSpeed() == 1);
    testData.reset();

    car.setTopSpeed(-1);
    writer.writeObject(car);
    carBack = reader.template readStatic<Car>();
    ASSERT(carBack.getTopSpeed() == -1);
    testData.reset();

}

BEGIN_TEST("Fixed point")

    ExtremeValuesTestData testData;
    ClassRegistry& registry = testData.registry;
    const ClassRegistry::EntryMap& entries = registry.entries();

    FOREACH_SERIALIZER(testFixedPoint);

END_TEST

/////////////////////////////////////////////////////////////////////

template<typename WriterType, typename ReaderType>
static void testFloatingPoint(ExtremeValuesTestData& testData, WriterType& writer, ReaderType& reader) {

    Car car, carBack;
    car._setAllFieldsSet(true, mgen::DEEP);

    car.getPositioningMutable().getVelocityMutable().setY(0.0f);
    writer.writeObject(car);
    carBack = reader.template readStatic<Car>();
    ASSERT(carBack.getPositioning().getVelocity().getY() == 0.0f);
    testData.reset();

    car.getPositioningMutable().getVelocityMutable().setY(1.5f);
    writer.writeObject(car);
    carBack = reader.template readStatic<Car>();
    ASSERT(FLOAT_APPRIX_EQ(carBack.getPositioning().getVelocity().getY(), 1.5f));
    testData.reset();

    car.getPositioningMutable().getVelocityMutable().setY(-1.5f);
    writer.writeObject(car);
    carBack = reader.template readStatic<Car>();
    ASSERT(FLOAT_APPRIX_EQ(carBack.getPositioning().getVelocity().getY(), -1.5f));
    testData.reset();

    car.getPositioningMutable().getVelocityMutable().setZ(0.0);
    writer.writeObject(car);
    carBack = reader.template readStatic<Car>();
    ASSERT(FLOAT_APPRIX_EQ(carBack.getPositioning().getVelocity().getZ(), 0.0));
    testData.reset();

    car.getPositioningMutable().getVelocityMutable().setZ(1.5);
    writer.writeObject(car);
    carBack = reader.template readStatic<Car>();
    ASSERT(FLOAT_APPRIX_EQ(carBack.getPositioning().getVelocity().getZ(), 1.5));
    testData.reset();

    car.getPositioningMutable().getVelocityMutable().setZ(-1.5);
    writer.writeObject(car);
    carBack = reader.template readStatic<Car>();
    ASSERT(FLOAT_APPRIX_EQ(carBack.getPositioning().getVelocity().getZ(), -1.5));
    testData.reset();

}

BEGIN_TEST("Floating point")

    ExtremeValuesTestData testData;
    ClassRegistry& registry = testData.registry;
    const ClassRegistry::EntryMap& entries = registry.entries();

    FOREACH_SERIALIZER(testFloatingPoint);

END_TEST

/////////////////////////////////////////////////////////////////////

END_TEST_GROUP

File: mgen-javalib/src/test/scala/se/culvertsoft/mgen/javapack/test/ExtremeValues.scala
package se.culvertsoft.mgen.javapack.test

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream

import scala.annotation.elidable
import scala.annotation.elidable.ASSERTION

import org.junit.Test

import gameworld.types.ClassRegistry
import gameworld.types.basemodule1.Car
import se.culvertsoft.mgen.javapack.metadata.FieldSetDepth
import se.culvertsoft.mgen.javapack.serialization.BinaryReader
import se.culvertsoft.mgen.javapack.serialization.BinaryWriter
import se.culvertsoft.mgen.javapack.serialization.BuiltInReader
import se.culvertsoft.mgen.javapack.serialization.BuiltInWriter
import se.culvertsoft.mgen.javapack.serialization.JsonPrettyWriter
import se.culvertsoft.mgen.javapack.serialization.JsonReader
import se.culvertsoft.mgen.javapack.serialization.JsonWriter

case class Precision(val p: Double)

object FloatsAlmostEqual {
  implicit class DoubleWithAlmostEquals(val d: Double) extends AnyVal {
    def ~=(d2: Double)(implicit p: Precision) = (d - d2).abs <= p.p
  }
}

class ExtremeValues {

  class TestState() {
    val classRegistry = new ClassRegistry
    val classRegEntries = classRegistry.entries()
    val stream = new ByteArrayOutputStream
    val jsonWriter = new JsonWriter(stream, classRegistry)
    val jsonWriterCompact = new JsonWriter(stream, classRegistry, true)
    val jsonPrettyWriter = new JsonPrettyWriter(stream, classRegistry)
    val jsonPrettyWriterCompact = new JsonPrettyWriter(stream, classRegistry, true)
    val binaryWriter = new BinaryWriter(stream, classRegistry)
    val binaryWriterCompact = new BinaryWriter(stream, classRegistry, true)

    val writers = Seq(
      jsonWriter,
      jsonWriterCompact,
      jsonPrettyWriter,
      jsonPrettyWriterCompact,
      binaryWriter,
      binaryWriterCompact)

    def reset() { stream.reset() }
  }

  def getReader(
    writer: BuiltInWriter)(implicit state: TestState): BuiltInReader = {
    writer match {
      case writer: JsonWriter =>
        new JsonReader(new ByteArrayInputStream(state.stream.toByteArray()), state.classRegistry)
      case writer: BinaryWriter =>
        new BinaryReader(new ByteArrayInputStream(state.stream.toByteArray()), state.classRegistry)
    }
  }

  @Test
  def testFixedPoint() {

    implicit val state = new TestState

    for (writer <- state.writers) {
      val car = new Car
      car._setAllFieldsSet(true, FieldSetDepth.DEEP)

      writer.writeObject(car.setId(Long.MaxValue))
      writer.writeObject(car.setId(Long.MinValue))
      writer.writeObject(car.setId(0))
      writer.writeObject(car.setId(1))
      writer.writeObject(car.setId(-1))

      writer.writeObject(car.setTopSpeed(Int.MaxValue))
      writer.writeObject(car.setTopSpeed(Int.MinValue))
      writer.writeObject(car.setTopSpeed(0))
      writer.writeObject(car.setTopSpeed(1))
      writer.writeObject(car.setTopSpeed(-1))

      val reader = getReader(writer)

      assert(reader.readObject(classOf[Car]).getId() == Long.MaxValue)
      assert(reader.readObject(classOf[Car]).getId() == Long.MinValue)
      assert(reader.readObject(classOf[Car]).getId() == 0)
      assert(reader.readObject(classOf[Car]).getId() == 1)
      assert(reader.readObject(classOf[Car]).getId() == -1)

      assert(reader.readObject(classOf[Car]).getTopSpeed() == Int.MaxValue)
      assert(reader.readObject(classOf[Car]).getTopSpeed() == Int.MinValue)
      assert(reader.readObject(classOf[Car]).getTopSpeed() == 0)
      assert(reader.readObject(classOf[Car]).getTopSpeed() == 1)
      assert(reader.readObject(classOf[Car]).getTopSpeed() == -1)

      state.reset

    }

  }

  @Test
  def testFloatingPoint() {
    import FloatsAlmostEqual._
    implicit val p = Precision(1e-5)

    implicit val state = new TestState

    for (writer <- state.writers) {
      val car = new Car
      car._setAllFieldsSet(true, FieldSetDepth.DEEP)

      car.getPositioning().getPosition().setY(0.0f)
      writer.writeObject(car)
      car.getPositioning().getPosition().setY(1.5f)
      writer.writeObject(car)
      car.getPositioning().getPosition().setY(-1.5f)
      writer.writeObject(car)
      car.getPositioning().getPosition().setZ(0.0)
      writer.writeObject(car)
      car.getPositioning().getPosition().setZ(1.5)
      writer.writeObject(car)
      car.getPositioning().getPosition().setZ(-1.5)
      writer.writeObject(car)

      val reader = getReader(writer)

      assert(reader.readObject(classOf[Car]).getPositioning().getPosition().getY() ~= -0.0f)
      assert(reader.readObject(classOf[Car]).getPositioning().getPosition().getY() ~= 1.5f)
      assert(reader.readObject(classOf[Car]).getPositioning().getPosition().getY() ~= -1.5f)

      assert(reader.readObject(classOf[Car]).getPositioning().getPosition().getZ() ~= -0.0)
      assert(reader.readObject(classOf[Car]).getPositioning().getPosition().getZ() ~= 1.5)
      assert(reader.readObject(classOf[Car]).getPositioning().getPosition().getZ() ~= -1.5)

      state.reset

    }

  }

}